sql/hive/src/main/scala/org/apache/spark/sql/hive/client/ClientWrapper.scala

```diff
@@ -62,6 +63,39 @@ private[hive] class ClientWrapper(
   extends ClientInterface
   with Logging {
 
+  overrideHadoopShims()
+
+  // !! HACK ALERT !!
+  //
+  // This method is used to workaround CDH Hadoop versions like 2.0.0-mr1-cdh4.1.1.
+  //
+  // Internally, Hive `ShimLoader` tries to load different versions of Hadoop shims by checking
+  // version information gathered from Hadoop jar files.  If the major version number is 1,
+  // `Hadoop20SShims` will be loaded.  Otherwise, if the major version number is 2, `Hadoop23Shims`
+  // will be chosen.  However, CDH Hadoop versions like 2.0.0-mr1-cdh4.1.1 have 2 as major version
+  // number, but contain Hadoop 1 code.  This confuses Hive `ShimLoader` and loads wrong version of
```

Comment:
> Is it OK to assume that all these versions should be used together with Hadoop20SShims, just like all the Hadoop 1.x.x versions?

No, that would be a wrong assumption. The first version in those strings is the Hadoop version the artifact is based on - so `2.3.0-mr1-cdh5.1.5` would contain, at the very least, the same API as Hadoop 2.3.0. So it should use the 2.3 shims. `mr1` here is irrelevant to identify the version of Hadoop.